packages/scrapers-v2/src/urls/scrape_urls.ts

```diff
@@ -0,0 +1,162 @@
+      path: [...paths, college],
+    }));
+    const catalogPaths = await scrapeLinks(baseUrl, initStack);
+    return convertToHierarchy(baseUrl, catalogPaths);
+  } catch (e) {
+    throw e;
+  }
+};
+
+/**
+ *
+ * Retrieves all sub-entries of the given initial queue in BFS fashion using the catalog sidebar hierarchy.
+ *  
+ * @param baseUrl   the base catalog URL, i.e. https://catalog.northeastern.edu
+ * @param initQueue a queue of parent entries
+ * @returns         a flat list of all the last level children catalog entries
+ */
+const scrapeLinks = async (
+  baseUrl: string,
+  initQueue: CatalogPath[]
+): Promise<CatalogPath[]> => {
+  const results: CatalogPath[] = [];
+
+  let queue = initQueue;
+  while (queue.length > 0) {
+    const pages = await Promise.all(getUrlHtmls(queue, baseUrl));
+    const nextQueue: CatalogPath[] = [];
+    for (const { $, url } of pages) {
+      const children = getChildrenForPathId($, url).toArray().map($);
+      for (const element of children) {
+        const path = getLinkForEl(element);
+        const bucket = isParent(element) ? nextQueue : results;
+        bucket.push({ path });
+      }
+    }
+    queue = nextQueue;
+  }
+
+  return results;
+};
+
+/**
+ * Converts a flat list of entries to catalog hierarchy.
+ * 
+ * @param base         the base catalog URL, i.e. https://catalog.northeastern.edu
+ * @param catalogPaths a flat list of paths
+ * @returns            catalog hierarchy
+ */
+const convertToHierarchy = (
+  base: string,
+  catalogPaths: CatalogPath[]
+): CatalogHierarchy => {
+  const hierarchy: CatalogHierarchy = {};
+  for (const { path } of catalogPaths) {
+    let obj: CatalogHierarchy = hierarchy;
+
+    // For each part of the path, add it to the hierarchy
+    for (let i = 0; i < path.length - 1; i += 1) {
+      const part = path[i];
+      if (!(part in obj)) {
+        obj[part] = {};
+      }
+      const child = obj[part];
+      if (typeof child === "string") {
+        throw new Error(
+          "Hierarchy was inconsistent: found a child, where a parent was expected"
+        );
+      }
+      obj = child;
+    }
+
+    const last = path[path.length - 1];
+    // Obj should equal the parent of the entry
+    obj[last] = joinParts(base, path).toString();
+  }
+  return hierarchy;
+};
+
+const isParent = (el: Cheerio) => {
+  return el.hasClass("isparent");
+};
+
+const getPathParts = (url: string) => {
+  return url.split("/").filter((s) => s !== "");
+};
+
+const getLinkForEl = (element: Cheerio): string[] => {
+  const aTag = element.find("a");
+  if (aTag.length === 0) {
+    const msg = "Catalog is missing a link for a parent element.";
+    throw new Error(msg);
+  }
+
+  return getPathParts(aTag.attr("href"));
+};
+
+const getChildrenForPathId = ($: CheerioStatic, url: URL) => {
+  // for getElementById, forward slashes need to be escaped
+  const id = url.pathname.split("/").join("\\/");
+  const current = $(`#${id}\\/`);
```

Comment:
add comment: The catalog entries have an ID equal to the path, with a trailing slash. We select the element via its ID